### PR TITLE
Utilize reportAction.clientGUID to prevent sequenceNumber mismatch

### PR DESCRIPTION
**Pull request by marcaaron**
### Fixed Issues
Fixes https://github.com/Expensify/Expensify/issues/141701

### Tests
1. Upload an attachment
1. Quickly leave a text comment
1. Verify that the text comment appears below the pending attachment
1. Verify that eventually the text comment moves above the pending attachment
1. Verify that one the attachment finishes uploading it will appear below the comment

### CODE REVIEW

**Review by tgolen — CHANGES_REQUESTED** [review]

**Review by cead22 — CHANGES_REQUESTED** [review]

File: src/libs/actions/Report.js
Comment (tgolen) [inline src/libs/actions/Report.js:60]: Can you please add to this comment to explain why it needs to be cleared, and why the recent fetch of report history is the best time to clear it?

File: src/libs/actions/Report.js
Comment (tgolen) [inline src/libs/actions/Report.js:213]: Does it matter if it's a local action or not? I don't think so. I would suggest just renaming this to `removeAction` since it can be used to remove any action ID.

File: src/libs/actions/Report.js
Comment (tgolen) [inline src/libs/actions/Report.js:224]: What's the difference between a "temporary" action and a "local" action (from the method above)?

File: src/libs/actions/Report.js
Comment (tgolen) [inline src/libs/actions/Report.js:228]:
```suggestion
    // Reset the temporary report action IDs to an empty array
```

File: src/libs/actions/Report.js
Comment (tgolen) [inline src/libs/actions/Report.js:261]: This comment refers to the "temporary" action, but it calls the method to remove a "local" action. Does the comment need updated or the method?

Also, why does the old one need to be removed if it's just going to get replaced? That essentially creates two writes to Onyx when only a single write is needed.

File: src/libs/actions/Report.js
Comment (tgolen) [inline src/libs/actions/Report.js:542]:
```suggestion
    const temporaryReportActionID = Str.guid(Date.now());
```

File: src/libs/actions/Report.js
Comment (tgolen) [inline src/libs/actions/Report.js:546]: This comment was hard for me to understand so I've added a suggestion, but I'm not entirely sure that it's right.
```suggestion
    // Store the temporary action ID on the report the comment was added to.
    // It will be removed later when refetching report actions in order to clear out any
    // stuck actions (ie. actions where the client never received a Pusher event, for 
    // whatever reason, from the server with the new action data).
```

File: src/pages/home/report/ReportActionPropTypes.js
Comment (tgolen) [reply src/pages/home/report/ReportActionPropTypes.js:16]: Should we maybe always cast these to a string?

**Review by tgolen — CHANGES_REQUESTED** [review]

File: src/libs/actions/Report.js
Comment (marcaaron) [reply src/libs/actions/Report.js:213]: I think mainly my reasoning was that we are in the actions file and it is less specific to say `removeAction` 
e.g. that could be confused with an API call that deletes an action. Not sure if that is important at this stage.

Comment (marcaaron) [reply src/libs/actions/Report.js:213]: `addAction()` calls the API `removeAction()` does not

File: src/libs/actions/Report.js
Comment (marcaaron) [reply src/libs/actions/Report.js:224]: A "local action" would be _any_ action in storage. The temporary actions are specifically actions that need to be paired with a server action or removed.

File: src/libs/actions/Report.js
Comment (marcaaron) [reply src/libs/actions/Report.js:261]: `removeLocalAction()` can remove any local action that we have a key for. In this case, it's a temporary action we are removing.

The old one needs to be removed because it does not have the same key as the action we are adding here. We use the `clientID` as the key/sequenceNumber for the temporary. The "real" action will just use its true `sequenceNumber` as a key.

Comment (marcaaron) [reply src/libs/actions/Report.js:261]: You're right about the two writes though .. right on!

Maybe we can do a single merge here instead and just do

```js
const actionsToMerge = {};
if (reportAction.clientID) {
    actionsToMerge[reportAction.clientID] = null;
}
actionsToMerge[reportAction.sequenceNumber] = {...};
Onyx.merge(`${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${reportID}`, actionsToMerge);
```

File: src/libs/actions/Report.js
Comment (marcaaron) [reply src/libs/actions/Report.js:546]: Great suggestion and much clearer thanks!

File: src/libs/actions/Report.js
Comment (marcaaron) [reply src/libs/actions/Report.js:213]: Maybe it would be clearer if this was `removeActionFromStorage()`

Comment (marcaaron) [reply src/libs/actions/Report.js:213]: actually I think we can just axe this and bulk remove everything

File: src/libs/actions/Report.js
Comment (tgolen) [reply src/libs/actions/Report.js:224]: OK, I think the term "local" can be a little confusing because I was thinking it was referring to something like a local variable. The temporary actions make sense, but maybe it would be better to refer to them specifically as `optimisticActions`?

For "local actions", maybe those can be referred to more specifically as "actions in storage" (looks like you suggested this above) or "actions in Onyx" (sux if we ever rename Onyx)?

File: src/libs/actions/Report.js
Comment (tgolen) [reply src/libs/actions/Report.js:261]: Oh, I like that, yeah.

File: src/libs/actions/Report.js
Comment (marcaaron) [reply src/libs/actions/Report.js:214]: That makes sense to me. Let me update any comments left over that are still referencing "local".

**Comment by marcaaron** [comment]
Updated. 

Re:  https://github.com/Expensify/Expensify.cash/pull/741#discussion_r549489290

>Should we maybe always cast these to a string?

Couldn't respond to this one for some reason, but I want to hold off on doing that as I'm not sure what the impacts would be.

**Review by tgolen — DISMISSED** [review]
Looks great, thanks!

**Comment by marcaaron** [comment]
Updated!

File: src/pages/home/report/ReportActionPropTypes.js
Comment (cead22) [inline src/pages/home/report/ReportActionPropTypes.js:16]: Sorry if I missed a convo, but can we make these always be strings?

File: src/libs/actions/Report.js
Comment (cead22) [inline src/libs/actions/Report.js:591]: NAB, but I'd love to stop truncating comments everywhere

**Review by cead22 — DISMISSED** [review]

File: src/libs/actions/Report.js
Comment (marcaaron) [reply src/libs/actions/Report.js:591]: Oh hmm sorry not sure what you mean by this? But happy to make any changes.

File: src/pages/home/report/ReportActionPropTypes.js
Comment (marcaaron) [reply src/pages/home/report/ReportActionPropTypes.js:16]: I'll have to look into that. But my first thought is that there are a lot of cases where they are not strings so they would need to be cast to strings only to make the `PropTypes` checker happy. Maybe that would resolve whatever concern there is here? And just to clarify... what is the concern?

File: src/pages/home/report/ReportActionPropTypes.js
Comment (marcaaron) [reply src/pages/home/report/ReportActionPropTypes.js:16]: Just curious, at what point should we cast them to a string? And why? I didn't do anything about this because I'm not too sure why it would be preferable to cast to a string, but maybe I'm missing something obvious.

**Comment by marcaaron** [comment]
Sorry this keeps falling off my radar. Not sure if there are any more comments to address here. Last time I checked we have approvals from @tgolen and @cead22. @MariaHCD did you want to take a look?

File: src/pages/home/report/ReportActionPropTypes.js
Comment (cead22) [reply src/pages/home/report/ReportActionPropTypes.js:16]: We discussed this on e.cash, imo this component should accept a single type for sequenceNumber

Comment (marcaaron) [reply src/pages/home/report/ReportActionPropTypes.js:16]: Yes, but I think the request here requires more consideration (I'll do my best to explain why). I agree we should solve it in a holistic way, but changing the `propTypes` here doesn't exactly achieve the desired outcome. For now, this accurately reflects what the data in Onyx can be and therefore accurately reflects the current state of things. Casting to a string would need to be done at a low level (in the Onyx data).

e.g. the `reportActions` here would need to have all `sequenceNumber` converted to strings in the Onyx data.

https://github.com/Expensify/Expensify.cash/blob/792277063ed8049952785112318ff670de5dbc31/src/pages/home/report/ReportActionsView.js#L24-L27

But if we coerce them to strings there could be unexpected results in other places (like `actions/Report.js`) where we expect `sequenceNumber` to be a number. We are also only using the `sequenceNumber` in the UI to create the React `key`.

I am less sure about casting all `sequenceNumber` to strings in storage and we can look into whether this is a safe change or not. But my preference would be to create a new issue and start sending string `sequenceNumber` at the API level if it makes sense to do that.

**Review by MariaHCD — DISMISSED** [review]
Looks good and tests well!

File: src/pages/home/report/ReportActionPropTypes.js
Comment (cead22) [reply src/pages/home/report/ReportActionPropTypes.js:16]: Thanks for the explanation.

Tbh, I'm not a huge fan, especially since this was accepting only numbers prior to this PR, so this feels like we're either introducing technical debt that we need to undo, or just adding on to existing one. 

That said, I don't want to block so go ahead with the new issue (and fwiw, I'd love for it to be high on your priority list)

Comment (marcaaron) [reply src/pages/home/report/ReportActionPropTypes.js:16]: Cool. Let's discuss more in the issue? https://github.com/Expensify/Expensify/issues/152504

File: src/libs/actions/Report.js
Comment (marcaaron) [reply src/libs/actions/Report.js:591]: Ah I think I figured out what you mean. We actually added a lint rule that requires us to break a line after a certain length. But sometimes we are breaking much earlier than that.

**Review by MariaHCD — APPROVED** [review]

**Pull request merged**
